pkg/multishare_reconciler/reconciler.go

```diff
@@ -0,0 +1,917 @@
+		return
+	}
+	klog.V(5).Infof("Found %d shares and %d instances", len(shares), len(instances))
+	instances, shares, instanceShares, err := recon.managedInstanceAndShare(instances, shares)
+	if err != nil {
+		klog.Errorf("Failed to filter out managed instance and shares: %s", err.Error())
+		return
+	}
+
+	// Create shareInfo objects if does not exist, update shareInfo.Status based on listed out shares' status.
+	shareInfoMap := recon.createAndUpdateShareInfos(shares)
+
+	shareInfoList, err := recon.shareLister.List(labels.Everything())
+	if err != nil {
+		klog.Errorf("Filestore CSI driver cannot list ShareInfo objects: %v", err)
+		return
+	}
+	klog.V(6).Infof("Listed out %d shareInfo objects", len(shareInfoList))
+	for _, shareInfo := range shareInfoList {
+		shareName := shareInfo.Name
+		if _, ok := shareInfoMap[shareName]; !ok {
+			shareInfoMap[shareName] = shareInfo
+		}
+	}
+
+	// Create instanceInfo objects if it does not exist, update instanceInfo.Status based on listed out instances' status.
+	instanceInfoMap := recon.createAndUpdateInstanceInfos(instances, instanceShares)
+
+	instanceInfoList, err := recon.instanceLister.List(labels.Everything())
+	if err != nil {
+		klog.Errorf("Filestore CSI driver cannot list InstanceInfo objects: %v", err)
+		return
+	}
+	klog.V(6).Infof("Listed out %d instanceInfo objects", len(instanceInfoList))
+	for _, instanceInfo := range instanceInfoList {
+		instanceURI := util.InstanceInfoNameToInstanceURI(instanceInfo.Name)
+		if _, ok := instanceInfoMap[instanceURI]; !ok {
+			instanceInfo, err := recon.maybeRemoveInstanceInfoFinalizer(instanceInfo)
+			if err != nil {
+				klog.Errorf("Error removing finalizer from InstanceInfo %q: %v", instanceInfo.Name, err)
+				continue
+			}
+			if instanceInfo != nil {
+				instanceInfoMap[instanceURI] = instanceInfo
+			}
+		}
+	}
+
+	// Assign un-assigned shares to instances; update shareInfo and instanceInfo accordingly,
+	// if there's inconsistency between share and instnace then share has source of truth.
+	recon.assignSharesToInstances(shareInfoMap, instanceInfoMap)
+
+	klog.Infof("Reconciliation round finished after %v", time.Since(startTime))
+}
+
+func (recon *multishareReconciler) assignSharesToInstances(shareInfos map[string]*v1alpha1.ShareInfo, instanceInfos map[string]*v1alpha1.InstanceInfo) {
+	recon.fixTwoWayPointers(shareInfos, instanceInfos)
+
+	recon.assignSharesToEligibleOrNewInstances(shareInfos, instanceInfos)
+
+	// Have to call deleteOrResizeInstances() after assigning shares and/or fixing two way pointers because no resizing were attempted in
+	// assignSharesToEligibleOrNewInstances() or fixTwoWayPointers()
+	recon.deleteOrResizeInstances(instanceInfos)
+}
+
+// fixTwoWayPointers scans over all instanceInfo objects and try to fix the 2 way pointer between instanceInfo and shareInfo objects.
+// instanceInfo.Status.ShareNames -> shareInfo
+// shareInfo.Status.InstanceHandle -> instanceInfo
```

Comment:
In the comment we should also mention the 2 way ptr fix prioritizes share parent reference and fixes the instance to point to the right share, and why

Reply:
done.